### PR TITLE
DownsampleAggregator with label

### PR DESCRIPTION
**Pull request by jackgoldsmith4**

### CODE REVIEW

**Comment by jackgoldsmith4** [comment]
Also change seqOpArgs in python to allow for additional seqOpArgs

File: python/hail/expr/aggregators/aggregators.py [hunk 1]
```diff
@@ -49,8 +49,8 @@ def __init__(self):
                       ret_type=hail_type,
                       constructor_args=sequenceof(expr_any),
                       init_op_args=nullable(sequenceof(expr_any)),
-                      f=nullable(func_spec(1, expr_any)))
-    def __call__(self, name, aggregable, ret_type, constructor_args=(), init_op_args=None, f=None):
+                      additional_seq_op_args=nullable(oneof(sequenceof(expr_any), func_spec(1, expr_any))))
+    def __call__(self, name, aggregable, ret_type, constructor_args=(), init_op_args=None, additional_seq_op_args=None):
```
Comment (cseed) [inline python/hail/expr/aggregators/aggregators.py]: OK, this code is a bit of a mess because the current aggregator design isn't quite right for multiple, but I think it can be improved.  First, `f` is terrible name and it's meaning is unclear.  I think we should have:

```
  def __call__(self, name, aggregable, ret_type, constructor_args=(), init_op_args=None, seq_op_args=()):
    ...
```

Then the seq_op and agg_sig can be constructed with:

```
        # no longer takes seq_op_args
        def agg_sig():
            return AggSignature(name,
                                list(map(get_type, constructor_args)),
                                None if init_op_args is None else list(map(get_type, init_op_args)),
                                list(map(get_type, seq_op_args)))

            def make_seq_op(agg):
                uid = Env.get_uid()
                ref = construct_variable(uid, get_type(agg), agg._indices)
                applied_seq_ops = [x(ref) if callable(x) else x for x in seq_op_args]
                return construct_expr(Let(uid, get_ir(agg), SeqOp(I32(0), [get_ir(x) for x in callable_seq_op_args], agg_sig())), None)
            seq_op = aggregable._transformations(aggregable, make_seq_op)
            signature = agg_sig()
```

Then we don't need to special case Count or test if `f` is `None`.  Any seq_op_arg can be an Expression or a callable that is expecting to take the aggregation value.  Count and the other aggregators that pass `f` will need to be modified slightly.

This is a bit gnarly so let's discuss in person if you have questions.

File: python/hail/expr/aggregators/aggregators.py [hunk 2]
```diff
@@ -1059,7 +1068,36 @@ def hist(expr, start, end, bins) -> StructExpression:
                 bin_freq=tarray(tint64),
                 n_smaller=tint64,
                 n_larger=tint64)
-    return _agg_func('Histogram', expr, t, [start, end, bins])
+    return _agg_func('Histogram', expr, t, constructor_args=[start, end, bins])
+
+
+@typecheck(x=expr_float64, y=expr_float64, label=nullable(oneof(expr_str, expr_array())), n_divisions=int)
```
Comment (cseed) [inline python/hail/expr/aggregators/aggregators.py]: This should be `expr_array(expr_str)`.

File: python/hail/plot/plots.py [hunk 3]
```diff
@@ -146,21 +147,36 @@ def scatter(x, y, label=None, title=None, xlabel=None, ylabel=None, size=4, lege
...
     legend : bool
         Whether or not to show the legend in the resulting figure.
+    collect_all : bool
+        Whether to collect all values or downsample before plotting.
+        Downsampling will only occur if x and y are both of type :class:`.Expression`.
+    n_divisions : int
+        Factor by which to downsample (default value = 500). A lower input results in fewer output datapoints.
     source_fields : Dict[str, List[Any]]
-        Extra fields for the ColumnDataSource of the plot.
+        Extra fields for the ColumnDataSource of the plot. Currently only used by calls from :meth:`manhattan`.
```
Comment (cseed) [inline python/hail/plot/plots.py]: I'd remove the comment it is only used by manhattan, because it is public, users could be using it, too!

File: python/hail/plot/plots.py [hunk 4]
```diff
@@ -146,21 +147,36 @@ def scatter(x, y, label=None, title=None, xlabel=None, ylabel=None, size=4, lege
         Size of markers in screen space units.
     legend : bool
         Whether or not to show the legend in the resulting figure.
+    collect_all : bool
```
Comment (cseed) [inline python/hail/plot/plots.py]: You might remark this is ignored if the input is a python object (e.g. it was already collected!)

File: python/hail/plot/plots.py [hunk 5]
```diff
@@ -317,7 +365,7 @@ def get_contig_index(x, starts):
             num_deleted += 1
 
     p = scatter(x, y, label=label, title=title, xlabel='Chromosome', ylabel='P-value (-log10 scale)',
-                size=size, legend=False, source_fields=hover_fields)
+                size=size, legend=False, source_fields=hover_fields, collect_all=True)
```
Comment (cseed) [inline python/hail/plot/plots.py]: No need to pass collect_all=True since it is ignored since x, y are Python objects.

File: src/main/scala/is/hail/stats/DownsampleCombiner.scala [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+  def buildFactor(coord: Double, lower: Double, upper: Double, factor: Int = 0): Int = {
+    assert(lower < upper)
+    if (coord < lower || coord > upper)
+      buildFactor(coord, lower - (upper - lower), upper + (upper - lower), factor + 1)
+    else
+      factor
+  }
+
+  def merge(x: Double, y: Double, label: Seq[String]): DownsampleCombiner = {
```
Comment (cseed) [inline src/main/scala/is/hail/stats/DownsampleCombiner.scala]: I think this can be improved as we discussed in person with baseMerge and a collapse function that uses the point to recompute the bins when collapsing.

File: python/hail/expr/aggregators/aggregators.py [hunk 7]
```diff
@@ -49,47 +49,44 @@ def __init__(self):
...
-                ref = construct_variable(uid, get_type(agg), agg._indices)
-                result = f(ref)
-                body = Let(uid, get_ir(agg), get_ir(result))
-                return construct_expr(SeqOp(I32(0), [get_ir(agg), body], agg_sig(agg, result)), None)
-            seq_op = aggregable._transformations(aggregable, make_seq_op)
-            signature = agg_sig(aggregable, f(aggregable))
+                                list(map(get_type, applied_seq_ops)))
+
+        def make_seq_op(agg):
+            ref = construct_variable(uid, get_type(agg), agg._indices)
```
Comment (cseed) [inline python/hail/expr/aggregators/aggregators.py]: I think you can move this into `apply_seq_ops` and eliminate the `ref` argument.

File: src/main/scala/is/hail/annotations/aggregators/RegionValueDownsampleAggregator.scala [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  
+  def combOp(agg2: RegionValueAggregator) = combiner.merge(agg2.asInstanceOf[RegionValueDownsampleAggregator].combiner)
+
+  def result(rvb: RegionValueBuilder) {
+    rvb.startArray(combiner.binToPoint.size)
+    combiner.binToPoint.foreach { case (_, (x, y, a)) =>
+      rvb.startTuple()
+      rvb.addDouble(x)
+      rvb.addDouble(y)
+      if (a != null) {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/aggregators/RegionValueDownsampleAggregator.scala]: If a is null, the array should be missing (not null).  You can do this by calling `rvb.addAnnotation(TArray(TString()), a)` and that will handle the case `a` is `null`.

File: src/main/scala/is/hail/stats/DownsampleCombiner.scala [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    } else
+      initPoint = (x, y, label)
+
+    this
+  }
+
+  def merge(that: DownsampleCombiner): DownsampleCombiner = {
+    if (that.binToPoint != null)
+      that.binToPoint.foreach { case (_, (x, y, l)) => merge(x, y, l) }
+    else
```
Comment (cseed) [inline src/main/scala/is/hail/stats/DownsampleCombiner.scala]: It is possible for a partition to have no points, in which case that.initPoint could be null.  Check that here.

File: src/main/scala/is/hail/stats/DownsampleCombiner.scala [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+      if (math.abs(yFactor) > 1) {
+        collapse = true
+        if (yFactor < 0)
+          bottom = y
+        else
+          top = y
+      }
+
+      if (collapse) {
+        val oldBinPoint = Map.empty ++ binToPoint
```
Comment (cseed) [inline src/main/scala/is/hail/stats/DownsampleCombiner.scala]: No need for the `Map.empty ++`.  That makes a copy, but you can just use the old binToPoint since you're creating a new one from scratch.

File: src/main/scala/is/hail/stats/DownsampleCombiner.scala [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+  var binToPoint: Map[BinType, PointType] = _
+  var initPoint: PointType = _
+
+  var left: Double = 0.0
+  var right: Double = 0.0
+  var bottom: Double = 0.0
+  var top: Double = 0.0
+
+  def toBinCoords(x: Double, y: Double, left: Double, right: Double, bottom: Double, top: Double, nDivisions: Int): (Int, Int) = {
```
Comment (cseed) [inline src/main/scala/is/hail/stats/DownsampleCombiner.scala]: It doesn't look like this needs left, right, bottom, top, nDivisions since those are all directly accessible from here.

**Comment by jackgoldsmith4** [comment]
@cseed  back to you

File: src/main/scala/is/hail/annotations/aggregators/RegionValueDownsampleAggregator.scala (hunk 8 above)
Comment (cseed) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueDownsampleAggregator.scala]: Looks like you didn't address thi sone.

Comment (jackgoldsmith4) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueDownsampleAggregator.scala]: whoops, misread your earlier comment. Fixing now.
